lib/plug/conn.ex

```diff
@@ -934,6 +934,49 @@ defmodule Plug.Conn do
   end
 
   @doc """
+  Push a resource to the client.
+
+  If the adapter does not support server push then this is a noop.
+  """
+  @spec push(t, String.t, Keyword.t) :: t
+  def push(conn, path, headers \\ []) do
+    adapter_push(conn, path, headers)
+    conn
+  end
+
+  @doc """
+  Push a resource to the client.
+
+  If the adapter does not support server push then this function will raise.
+  """
+  @spec push!(t, String.t, Keyword.t) :: t
+  def push!(%Conn{adapter: {adapter, _}} = conn, path, headers \\ []) do
+    case adapter_push(conn, path, headers) do
+      :ok ->
+        conn
+
+      _ ->
+        raise "server push not supported by #{inspect(adapter)}." <>
+        "You should either delete the call to `push!/3` or switch to an " <>
+        "adapter that does support server push such as Plug.Adapters.Cowboy2."
+    end
+  end
+
+  defp adapter_push(%Conn{state: state}, _path, _headers)
+  when not state in @unsent do
+    raise AlreadySentError
```

Comment:
@josevalim Not sure if this error is appropriate or not. Whilst the fact a response is already sent is the cause of the error, it may not be clear to users why this is the cause.

As an aside, I think this fact should be mentioned in the docs, I'll add it in.

Reply:
This is great as is IMO.